# TB MODEL FOR MEDICAL PROGNOSIS

The project aims to create a clustering model for TB prognosis. The clustering model will assist in detecting the liklihood of TB acquisition based on the various parameters depicted.

C_1: 
Import all the necessary packages required for the project

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import sklearn

C_2: 
Import the data that already exists in csv form and exists in the env
Exp: low_memory=False enables pandas to read the data as a whole first before checking on the data set within the columns.
setting low_memory=True makes pandas to read the data in chunks and categorizes the data frame before proceeding to integrate the data inorder to create the dataframe

In [2]:
df=pd.read_csv('Tb data.csv',low_memory=False)

## 1.0 Exploratory Data Analysis 

Exploratory data anlysis involves a set of statistical component around data treatement with a focus on determining the general structure and achitecture of the dataframe: In the section we explore a few concepts of EDA
    1. General outlook of the data
    
    2. Missing data
    
    3. Data Imputation
    
    4. Duplicate data
    
    4. Handling outliers
    
    5. Bivariate Analysis
I have developed a norm to create copies of the original data set inorder to try out all the exploratory data analysis techniques
    

### 1.1 General Outlook of the data

The section mainly aims to show how to look around various aspects of the data set. This enables you to become more farmiliar and confortable with the data.
Inorder to read the all columns so that we may be in a position to check all properties we use the max columns option

In [3]:
pd.options.display.max_columns = None
df.head(15)

Sex  Age MoreThan30Days Eligible  q_101  q_103  q_107  q_108      q_109  \
0    M   47            Yes      Yes    NaN    0.0    0.0    2.0  722674159   
1    M   18            Yes      Yes    NaN    0.0    0.0    2.0        NaN   
2    F   46            Yes      Yes    NaN    0.0    0.0    2.0  710568882   
3    F   27            Yes      Yes    NaN    0.0    0.0    2.0  729580400   
4    M   31            Yes      Yes    NaN    0.0    0.0    2.0        NaN   
5    F    7            Yes       No    NaN    0.0    0.0    2.0        NaN   
6    F    4            Yes       No    NaN    0.0    0.0    2.0        NaN   
7    F   18            Yes      Yes    NaN    0.0    0.0    2.0  724699001   
8    M   25            Yes      Yes    NaN    0.0    0.0    2.0        NaN   
9    F    0            Yes       No    NaN    0.0    0.0    2.0        NaN   
10   M   33            Yes      Yes    NaN    0.0    0.0    2.0  724756524   
11   F   52            Yes      Yes    NaN    0.0    0.0    2.0  716968226   
12   M   54            Yes      Yes    NaN    0.0    0.0    2.0        NaN   
13   F   26            Yes      Yes    NaN    0.0    0.0    2.0  708863031   
14   M   30            Yes      Yes    NaN    0.0    0.0    2.0        NaN   

    q_111 DateStartedMFSParticipation  IsPregnant  SputumRequestDone  \
0     1.0      20JUN2016:14:13:18.000         0.0                NaN   
1     2.0      17JUN2016:15:42:05.000         0.0                NaN   
2     1.0      18JUN2016:16:50:07.000         0.0                NaN   
3     1.0      18JUN2016:10:48:01.000         0.0                NaN   
4     2.0                         NaN         NaN                NaN   
5     3.0                         NaN         NaN                NaN   
6     4.0                         NaN         NaN                NaN   
7     1.0      23JUN2016:18:48:07.000         0.0                NaN   
8     2.0      18JUN2016:16:19:21.000         0.0                NaN   
9     3.0                         NaN         NaN                NaN   
10    1.0      23JUN2016:12:46:15.000         0.0                NaN   
11    1.0      16JUN2016:17:33:46.000         0.0                NaN   
12    2.0                         NaN         NaN                NaN   
13    1.0      18JUN2016:18:00:02.000         0.0                NaN   
14    2.0      19JUN2016:12:31:07.000         0.0                NaN   

    FirstSputumCollected  SecondSputumCollected  XRayImageTaken  \
0                    NaN                    NaN             1.0   
1                    NaN                    NaN             1.0   
2                    NaN                    NaN             1.0   
3                    NaN                    NaN             1.0   
4                    NaN                    NaN             NaN   
5                    NaN                    NaN             NaN   
6                    NaN                    NaN             NaN   
7                    NaN                    NaN             1.0   
8                    NaN                    NaN             1.0   
9                    NaN                    NaN             NaN   
10                   NaN                    NaN             1.0   
11                   NaN                    NaN             1.0   
12                   NaN                    NaN             NaN   
13                   NaN                    NaN             1.0   
14                   NaN                    NaN             1.0   

    XRayImageAnalysisDone  IsTBSuspect  ParticipantCareSeekingDone  \
0                     NaN          NaN                         1.0   
1                     NaN          NaN                         1.0   
2                     NaN          NaN                         1.0   
3                     NaN          NaN                         1.0   
4                     NaN          NaN                         NaN   
5                     NaN          NaN                         NaN   
6                     NaN          NaN

In [ ]:
#print only one column to view the items
#print(df_tb.AntiTbTreatment.to_string(index=False))

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Columns: 269 entries, Sex to HSBDone
dtypes: float64(186), int64(3), object(80)
memory usage: 259.4+ MB


In [5]:
#Convert the two columns to date and time
df['TreatmentToDate'] =  pd.to_datetime(df['TreatmentToDate'], format='%d/%m/%Y')
df['TreatmentFromDate'] =  pd.to_datetime(df['TreatmentFromDate'], format='%d/%m/%Y')

In [ ]:
# #Creating new calculated field for anti-TB treatment
# df_tb['AntiTbTreatment']=df_tb['TreatmentToDate']-df_tb['TreatmentFromDate']
# df_tb.head(30)

In [6]:
#Check the information and shape of the data frame
df.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Data columns (total 269 columns):
 #   Column                            Dtype         
---  ------                            -----         
 0   Sex                               object        
 1   Age                               int64         
 2   MoreThan30Days                    object        
 3   Eligible                          object        
 4   q_101                             float64       
 5   q_103                             float64       
 6   q_107                             float64       
 7   q_108                             float64       
 8   q_109                             object        
 9   q_111                             float64       
 10  DateStartedMFSParticipation       object        
 11  IsPregnant                        float64       
 12  SputumRequestDone                 float64       
 13  FirstSputumCollected              float64       
 14  SecondSputumCollect

O_1: 
From the outlook of the data we observe that there are a few of the columns that contain missing data. The columns are represnted by NaN. The data also has a missed set of data types<>.

O_2:
The data contains a total of 126388 observations and 269 variables. The observations represent the number of rows in the dataframe and the variables represent the number of columns in the dataframe.
Further the dataframe contains 3 data types float64 is in 186 columns

Some of the datatypes may be wrongly labelled in pandas. Pandas makes assumption in uploading the data based on what is close to it. Howevr there are various known ways to convert the data types inorder to set the right data types with pandas <Convert data types in pandas> Some of the three main techniques are; Use astype() which forces an appropriate data set, Creation of a custom function to convert the data and pandas functions such as to_numeric() or to_date(time)

#### 1 Dropping the Irrelevant columns

In [8]:
df_dropcol=df.copy()
df_dropcol.head()

Sex  Age MoreThan30Days Eligible  q_101  q_103  q_107  q_108      q_109  \
0   M   47            Yes      Yes    NaN    0.0    0.0    2.0  722674159   
1   M   18            Yes      Yes    NaN    0.0    0.0    2.0        NaN   
2   F   46            Yes      Yes    NaN    0.0    0.0    2.0  710568882   
3   F   27            Yes      Yes    NaN    0.0    0.0    2.0  729580400   
4   M   31            Yes      Yes    NaN    0.0    0.0    2.0        NaN   

   q_111 DateStartedMFSParticipation  IsPregnant  SputumRequestDone  \
0    1.0      20JUN2016:14:13:18.000         0.0                NaN   
1    2.0      17JUN2016:15:42:05.000         0.0                NaN   
2    1.0      18JUN2016:16:50:07.000         0.0                NaN   
3    1.0      18JUN2016:10:48:01.000         0.0                NaN   
4    2.0                         NaN         NaN                NaN   

   FirstSputumCollected  SecondSputumCollected  XRayImageTaken  \
0                   NaN                    NaN             1.0   
1                   NaN                    NaN             1.0   
2                   NaN                    NaN             1.0   
3                   NaN                    NaN             1.0   
4                   NaN                    NaN             NaN   

   XRayImageAnalysisDone  IsTBSuspect  ParticipantCareSeekingDone  \
0                    NaN          NaN                         1.0   
1                    NaN          NaN                         1.0   
2                    NaN          NaN                         1.0   
3                    NaN          NaN                         1.0   
4                    NaN          NaN                         NaN   

   IsTakingTBDrugs  SendToSupervisor  SecondSputumRequestDone  \
0              NaN               NaN                      NaN   
1              NaN               NaN                      NaN   
2              NaN               NaN                      NaN   
3              NaN               NaN                      NaN   
4              NaN               NaN                      NaN   

   SentToCareSeekingBy DateSentToCareSeeking  NationalImagingAnalysisDone  \
0                  NaN                   NaN                          NaN   
1                  NaN                   NaN                          NaN   
2                  NaN                   NaN                          NaN   
3                  NaN                   NaN                          NaN   
4                  NaN                   NaN                          NaN   

  DateNationalImagingAnalysisDone  AgeGroup  Eligibles  Schooling  \
0                             NaN         1          3        5.0   
1                             NaN         1          3        5.0   
2                             NaN         1          3        1.0   
3                             NaN         1          3        5.0   
4                             NaN         1          3        NaN   

   MaritalStatus  Coughing  CoughingWeeksNumber  Sputum  SputumWeeksNumber  \
0            2.0       0.0                  0.0     0.0                0.0   
1            1.0       0.0                  0.0     0.0                0.0   
2            1.0       0.0                  0.0     0.0                0.0   
3            2.0       0.0                  0.0     0.0                0.0   
4            NaN       NaN                  NaN     NaN                NaN   

   BloodCough  BloodCoughWeeksNumber  ChestPains  ChestPainsWeeksNumber  \
0         0.0                    0.0         0.0                    0.0   
1         0.0                    0.0         0.0                    0.0   
2         0.0                    0.0         0.0                    0.0   
3         0.0                    0.0         0.0                    0.0   
4         NaN                    NaN         NaN                    NaN   

   Fever  FeverWeeksNumber  Fatigue  FatigueWeeksNumber  WeightLoss  \
0    0.0               0.0      0.0     

In [9]:
#If the column appear to contain insensible categories I delete
df_dropcol.drop(['DateNationalImagingAnalysisDone','DateStartedMFSParticipation',
            'DateStartedMFSParticipation','DateSentToCareSeeking',
            'NationalImagingAnalysisDone','DateNationalImagingAnalysisDone',
            'SputumWeeksNumber','BloodCoughWeeksNumber',
            'ChestPainsWeeksNumber','FeverWeeksNumber','FatigueWeeksNumber',
            'OtherSymptomsWeeksNumber','OtherSymptomsText','TreatmentFromDate',
            'TreatmentToDate','NightSweatsWeeksNumber',
            'FacilityType','TreatmentWeeksNumber',
            'AccomodationCosts','FriendsSupportSource3',
            'SoldHouseAssetSource1','SoldHouseAssetSource2',
            'SoldHouseAssetSource3','NHIFSource1','NHIFSource2',
            'NHIFSource3','BorrowedMoneySource1','BorrowedMoneySource2',
            'BorrowedMoneySource3','CreditSource2','CreditSource3',
            'DiagnosisCost1','DiagnosisCost2','DiagnosisCost3',
            'CheckupCost1','CheckupCost2','CheckupCost3',
            'ConsultationCost1','ConsultationCost2',
            'ConsultationCost3','TotalCostsVisit1','TotalCostsVisit2',
            'TotalCostsVisit3','TreatmentExpenditureDontKnowVis1',
            'TreatmentExpenditureDontKnowVisi',
            'TreatmentExpenditureDontKnowVis0','TotalCostsVisit1',
 'SourceOfFundsFroTreatment1','SourceOfFundsFroTreatment2','Country',
            'DistrictHospital','DHospitalWeeksNumber','PrivatePractitioner',
            'PrivatePractitionerWeeksNumber', 'OtherHealthProviderWeeksNumber',
            'HIVStattusKnow','HIVStatusDisclose',
            'OtherCurrenTreatmentHealthCenter','CurrentTreatmentFromDate',
            'TBTreatmentCardNumber','SOPInstanceUID','idealread_4',
            'InterpretationField','ContainerLabelGenerated_1',
            'ContainerLabelGeneratedBy_1','DateContainerLabelGenerated_1',
            'DateDone_1','DoneBy_1','SputumCollectedDate_1','resposibility_1',
            'ResposibilityDate_1','ContainerLabelGenerated_2',
            'ContainerLabelGeneratedBy_2','DateContainerLabelGenerated_2',
            'RequestSequence_2','SmearCheck_2','SputumCollectedDate_2',
            'sample_number','sample_type','Smear','MTB','transit_time',
            'mfs_barcode','xpert_date','colony','review_date','ipa',
            'smearDate','CultureDate','DateOfSampleReception','TempResult',
            'sample_number_2','Smear_2','RIF_2','submission_date_2',
            'transit_time_2','DateInt_2','volume_2','review_date_2','ipa_2',
           ], axis=1, inplace=True)

In [10]:
#Data set ater dropping the irrelevant features to the data
df_dropcol.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Data columns (total 172 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Sex                               object 
 1   Age                               int64  
 2   MoreThan30Days                    object 
 3   Eligible                          object 
 4   q_101                             float64
 5   q_103                             float64
 6   q_107                             float64
 7   q_108                             float64
 8   q_109                             object 
 9   q_111                             float64
 10  IsPregnant                        float64
 11  SputumRequestDone                 float64
 12  FirstSputumCollected              float64
 13  SecondSputumCollected             float64
 14  XRayImageTaken                    float64
 15  XRayImageAnalysisDone             float64
 16  IsTBSuspect                       flo

In [11]:
df_dropcol.to_csv(r'C:\Users\Rogers\export_dftb.csv')

In [13]:
#check the unique caregories of a variable
print (df_dropcol['Culturepositive'].unique())

[nan 'NEG' 'POS' 'CON']


In [14]:
categories = df_dropcol.select_dtypes(exclude=['int', 'float'])
for col in df_dropcol.columns:
    #print(df_tb_categories[col].unique()) # to print categories name only
    print(df_dropcol['Culturepositive'].value_counts()) # to print count of every category


NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype

NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype: int64
NEG    9424
CON     385
POS     225
Name: Culturepositive, dtype

There are some columns which will not provide so much information to the model and therefore we may drop the variables at this moment. Some of those variable incude the 'DateNationalImagingAnalysisDone' which has shown to have no categories 
Res: There occures to be incomplete data from that displayed in the questionnare. THe data mainly is on an household social econoic score.
The variable q109 appears to b corrupt. should be yes/No

How do we deal with the text data inside the data OtherSymptomsText, Train a dictionary to remove the health facilities

#### Reordering the position of the columns in the dataset

#The data set has three response variables namely smear positive, culture positive and expert positive. We now move the response variables to the tail end of the data. https://towardsdatascience.com/reordering-pandas-dataframe-columns-thumbs-down-on-standard-solutions-1ff0bc2941d5

In [15]:
df_reorder_response=df_dropcol.copy()

In [16]:
#pop the feaature smearpositive from the dataset
col_name="Smearpositive"
smear_col = df_reorder_response.pop(col_name)

In [17]:
#pop the feaature Xpertpositive from the dataset
col_namex="Xpertpositive"
xpert_col = df_reorder_response.pop(col_namex)

In [18]:
#pop the feaature culturepositive from the dataset
col_namec="Culturepositive"
culture_col = df_reorder_response.pop(col_namec)

In [19]:
df_reorder_response.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Data columns (total 169 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Sex                               object 
 1   Age                               int64  
 2   MoreThan30Days                    object 
 3   Eligible                          object 
 4   q_101                             float64
 5   q_103                             float64
 6   q_107                             float64
 7   q_108                             float64
 8   q_109                             object 
 9   q_111                             float64
 10  IsPregnant                        float64
 11  SputumRequestDone                 float64
 12  FirstSputumCollected              float64
 13  SecondSputumCollected             float64
 14  XRayImageTaken                    float64
 15  XRayImageAnalysisDone             float64
 16  IsTBSuspect                       flo

In [20]:
df_reorder_response.insert(169,'Smearpositive', smear_col)

In [21]:
df_reorder_response.insert(170,'Xpertpositive', xpert_col)

In [22]:
df_reorder_response.insert(171,'Culturepositive', culture_col)

In [23]:
df_reorder_response.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Data columns (total 172 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Sex                               object 
 1   Age                               int64  
 2   MoreThan30Days                    object 
 3   Eligible                          object 
 4   q_101                             float64
 5   q_103                             float64
 6   q_107                             float64
 7   q_108                             float64
 8   q_109                             object 
 9   q_111                             float64
 10  IsPregnant                        float64
 11  SputumRequestDone                 float64
 12  FirstSputumCollected              float64
 13  SecondSputumCollected             float64
 14  XRayImageTaken                    float64
 15  XRayImageAnalysisDone             float64
 16  IsTBSuspect                       flo

## 2.0 Missing data

#### Checking on the percentage proportion of Missing data

In [25]:
df_missing_percentage=df_reorder_response.copy()

In [27]:
#determinig the proportion of the missing data to the overall data set
total = df_missing_percentage.isnull().sum().sort_values(ascending=False)
percent = (df_missing_percentage.isnull().sum()/df_missing_percentage.shape[0]).sort_values(ascending=False)

missing = pd.concat([total, percent], axis=1, keys=['Total', 'Percentage_missing'])
missing.head(50)

Total  Percentage_missing
q_101                             126389            1.000000
CreationTime                      126389            1.000000
AP                                126389            1.000000
OtherHealthServiceProvider        126383            0.999953
i_2                               126258            0.998964
r_2                               126258            0.998964
s_2                               126258            0.998964
e_2                               126258            0.998964
i                                 126247            0.998876
r                                 126247            0.998876
e                                 126247            0.998876
s                                 126245            0.998861
colony_2                          126142            0.998046
ica_2                             126139            0.998022
mtbresult_2                       126086            0.997603
MTB_2                             126086            0.997603
DidNotUsePublicHealthOtherReason  126042            0.997255
ica                               126042            0.997255
IsTBSuspect                       125617            0.993892
IsTakingTBDrugs                   125617            0.993892
SentToCareSeekingBy               125617            0.993892
XRayImageAnalysisDone             125603            0.993781
TreatmentHealthCenter             125487            0.992863
ClinicianType_4                   125464            0.992681
Interpretation_4                  125464            0.992681
read_4                            125464            0.992681
SecondSputumRequestDone           125305            0.991423
quality_2                         121943            0.964823
quality                           121446            0.960891
SendToSupervisor                  121420            0.960685
cultureresult_2                   117944            0.933182
smearresult_2                     117944            0.933182
Culture_2                         117944            0.933182
sample_type_2                     117942            0.933167
TempResult_2                      117942            0.933167
recieve_date_2                    117942            0.933167
mfs_barcode_2                     117942            0.933167
rep_2                             117942            0.933167
status_2                          117942            0.933167
samplemorning                     117724            0.931442
RIF                               117438            0.929179
SputumCollected_2                 117348            0.928467
samplespot                        116572            0.922327
mtbresult                         116571            0.922319
Xpertpositive                     116567            0.922288
MTBs                              116567            0.922288
submission_date                   116426            0.921172
smearresult                       116357            0.920626
Culturepositive                   116355            0.920610
smears                            116355            0.920610

Missing variables provide a very difficult situation to handle. The most common way will be dropping the variables and remaining with the columns that have variables. However for the case of this experiment it may result in dropping over 95% of the variables which would result in loosing alot of information. Therefore I try out a few imputation techniques.

#### 2 Dropping the columns with completely missing data

In [31]:
data=df_missing_percentage.drop((missing[missing['Percentage_missing'] >= 1.0]).index,1)
data.head(15)

Sex  Age MoreThan30Days Eligible  q_103  q_107  q_108      q_109  q_111  \
0    M   47            Yes      Yes    0.0    0.0    2.0  722674159    1.0   
1    M   18            Yes      Yes    0.0    0.0    2.0        NaN    2.0   
2    F   46            Yes      Yes    0.0    0.0    2.0  710568882    1.0   
3    F   27            Yes      Yes    0.0    0.0    2.0  729580400    1.0   
4    M   31            Yes      Yes    0.0    0.0    2.0        NaN    2.0   
5    F    7            Yes       No    0.0    0.0    2.0        NaN    3.0   
6    F    4            Yes       No    0.0    0.0    2.0        NaN    4.0   
7    F   18            Yes      Yes    0.0    0.0    2.0  724699001    1.0   
8    M   25            Yes      Yes    0.0    0.0    2.0        NaN    2.0   
9    F    0            Yes       No    0.0    0.0    2.0        NaN    3.0   
10   M   33            Yes      Yes    0.0    0.0    2.0  724756524    1.0   
11   F   52            Yes      Yes    0.0    0.0    2.0  716968226    1.0   
12   M   54            Yes      Yes    0.0    0.0    2.0        NaN    2.0   
13   F   26            Yes      Yes    0.0    0.0    2.0  708863031    1.0   
14   M   30            Yes      Yes    0.0    0.0    2.0        NaN    2.0   

    IsPregnant  SputumRequestDone  FirstSputumCollected  \
0          0.0                NaN                   NaN   
1          0.0                NaN                   NaN   
2          0.0                NaN                   NaN   
3          0.0                NaN                   NaN   
4          NaN                NaN                   NaN   
5          NaN                NaN                   NaN   
6          NaN                NaN                   NaN   
7          0.0                NaN                   NaN   
8          0.0                NaN                   NaN   
9          NaN                NaN                   NaN   
10         0.0                NaN                   NaN   
11         0.0                NaN                   NaN   
12         NaN                NaN                   NaN   
13         0.0                NaN                   NaN   
14         0.0                NaN                   NaN   

    SecondSputumCollected  XRayImageTaken  XRayImageAnalysisDone  IsTBSuspect  \
0                     NaN             1.0                    NaN          NaN   
1                     NaN             1.0                    NaN          NaN   
2                     NaN             1.0                    NaN          NaN   
3                     NaN             1.0                    NaN          NaN   
4                     NaN             NaN                    NaN          NaN   
5                     NaN             NaN                    NaN          NaN   
6                     NaN             NaN                    NaN          NaN   
7                     NaN             1.0                    NaN          NaN   
8                     NaN             1.0                    NaN          NaN   
9                     NaN             NaN                    NaN          NaN   
10                    NaN             1.0                    NaN          NaN   
11                    NaN             1.0                    NaN          NaN   
12                    NaN             NaN                    NaN          NaN   
13                    NaN             1.0                    NaN          NaN   
14                    NaN             1.0                    NaN          NaN   

    ParticipantCareSeekingDone  IsTakingTBDrugs  SendToSupervisor  \
0                          1.0              NaN               NaN   
1                          1.0              NaN               NaN   
2                          1.0              NaN               NaN   
3                          1.0              NaN               NaN   
4                          NaN              NaN               NaN   
5                          NaN              NaN               NaN   
6                          Na

In [32]:
#Checking the data frame after dropping some irrelevant columns
data.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126389 entries, 0 to 126388
Data columns (total 169 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Sex                               object 
 1   Age                               int64  
 2   MoreThan30Days                    object 
 3   Eligible                          object 
 4   q_103                             float64
 5   q_107                             float64
 6   q_108                             float64
 7   q_109                             object 
 8   q_111                             float64
 9   IsPregnant                        float64
 10  SputumRequestDone                 float64
 11  FirstSputumCollected              float64
 12  SecondSputumCollected             float64
 13  XRayImageTaken                    float64
 14  XRayImageAnalysisDone             float64
 15  IsTBSuspect                       float64
 16  ParticipantCareSeekingDone        flo

### Handling imbalanced data
There may be a possibility that the data on tb has an imbalanced data set in some of the variables. Let say the features within a data set have very many observations relating to one sample.

The data set on TB has an imbalanced data set because most of the observations do not have tb.

In [ ]:
df_tb_imbalance=df_tb.copy()
df_tb_imbalance.head()

#### Undersampling

In [ ]:
#Create independent and dependent features
columns=df_tb_imbalance.columns.tolist()
#filter the columns to remove the data we don't want
columns=[c for c in columns if c not in ["TBSuspect"]]
#Store the variable we are predicting
target="TBSuspect"
#Define a random state
state = np.random.RandomState(42)
X=df_tb_imbalance
Y = df_tb_imbalance[target]
X_outliers = state.uniform(low=0, high=1, size=(X.shape[0], X.shape[1]))
# Print the shapes of X & Y
print(X.shape)
print(Y.shape)

In [ ]:
#Check if there are any available 
df_tb_imbalance.isnull().values.any()

In [ ]:
count_tbsuspect = pd.value_counts(df_tb_imbalance['TBSuspect'], sort = True)

count_tbsuspect.plot(kind = 'bar', rot=0)

plt.title("TB Suspect distribution")

plt.xticks(range(2), )

plt.xlabel("TBSuspect")

plt.ylabel("Frequency")

In [ ]:
#check the unique caregories of a variable
print (df_tb_imbalance['TBSuspect'].unique())

In [ ]:
df_tb_imbalance.shape

In [ ]:
## Get the Fraud and the normal dataset 

positive = df_tb_imbalance[df_tb_imbalance['TBSuspect']=='POS']

negative = df_tb_imbalance[df_tb_imbalance['TBSuspect']=='NEG']

unidentified = df_tb_imbalance[df_tb_imbalance['TBSuspect']=='NON']

notapplicable = df_tb_imbalance[df_tb_imbalance['TBSuspect']=='nan']

In [ ]:
print(positive.shape,negative.shape,unidentified.shape,notapplicable.shape)

O_3: The data indicates that there is an aspect of inbalanced data set has more sampes in the negative form than the positive form
The second question however begs that the entire data set consists of 126389 but total of classification for both positives and negatives indicates only a total 

In [ ]:
#Due to the high number 

In [ ]:
count_AllLabResult = pd.value_counts(df_tb_imbalance['AllLabResult'], sort = True)

count_AllLabResult.plot(kind = 'bar', rot=0)

plt.title("TB Suspect distribution")

plt.xticks(range(2), )

plt.xlabel("AllLabResult ")

plt.ylabel("Frequency")

In [ ]:
from imblearn.under_sampling import NearMiss

In [ ]:

# Implementing Undersampling for Handling Imbalanced 
nm = NearMiss(random_state=42)
X_res,y_res=nm.fit_sample(X,Y)

In [ ]:
X_res.shape,y_res.shape


In [ ]:
from collections import Counter
print('Original dataset shape {}'.format(Counter(Y)))
print('Resampled dataset shape {}'.format(Counter(y_res)))

In [ ]:
import numpy as np
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='mean')
imp.fit([[1, 2], [np.nan, 3], [7, 6]])
SimpleImputer()
X = [[np.nan, 210], [400, np.nan], [700, 6]]
print(imp.transform(X))

In [ ]:
import scipy.sparse as sp
X = sp.csc_matrix([[1, 2], [0, -1], [8, 4]])

In [ ]:
import sklearn
import numpy as np

In [ ]:
from sklearn.datasets.california_housing import fetch_california_housing

In [ ]:
[dataset for dataset in sklearn.datasets.__dict__.keys() if str(dataset).startswith("fetch_") or str(dataset).startswith("load_")]


In [ ]:
houses = fetch_california_housing()

In [ ]:
print(houses.DESCR)

In [ ]:
import pandas as pd


In [ ]:
df = pd.DataFrame(houses.data, columns=houses.feature_names)
df.head(20)